src/main/java/com/transferwise/openbanking/client/api/vrp/RestVrpClient.java

```diff
@@ -0,0 +1,168 @@
+package com.transferwise.openbanking.client.api.vrp;
+
+import com.transferwise.openbanking.client.api.common.AuthorizationContext;
+import com.transferwise.openbanking.client.api.common.BasePaymentClient;
+import com.transferwise.openbanking.client.api.common.IdempotencyKeyGenerator;
+import com.transferwise.openbanking.client.api.common.OpenBankingHeaders;
+import com.transferwise.openbanking.client.api.payment.v3.model.vrp.OBDomesticVRPConsentRequest;
+import com.transferwise.openbanking.client.api.payment.v3.model.vrp.OBDomesticVRPDetails;
+import com.transferwise.openbanking.client.api.payment.v3.model.vrp.OBDomesticVRPRequest;
+import com.transferwise.openbanking.client.api.payment.v3.model.vrp.OBDomesticVRPResponse;
+import com.transferwise.openbanking.client.configuration.AspspDetails;
+import com.transferwise.openbanking.client.configuration.SoftwareStatementDetails;
+import com.transferwise.openbanking.client.error.ApiCallException;
+import com.transferwise.openbanking.client.json.JsonConverter;
+import com.transferwise.openbanking.client.jwt.JwtClaimsSigner;
+import com.transferwise.openbanking.client.oauth.OAuthClient;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+import org.springframework.web.client.RestOperations;
+
+@Slf4j
+public class RestVrpClient extends BasePaymentClient implements VrpClient {
+
+    private static final String VRP_RESOURCE = "domestic-vrps";
+    private static final String BASE_ENDPOINT_PATH_FORMAT = "%s/open-banking/v3.%s/vrp/%s";
+    private static final String VRP_BY_ID_ENDPOINT_PATH_FORMAT = BASE_ENDPOINT_PATH_FORMAT + "/{domesticVrpId}";
+    private static final String VRP_DETAILS_BY_ID_ENDPOINT_PATH_FORMAT = BASE_ENDPOINT_PATH_FORMAT + "/{domesticVrpId}/payment-details";
+
+
+    private final IdempotencyKeyGenerator<OBDomesticVRPConsentRequest, OBDomesticVRPRequest> idempotencyKeyGenerator;
+    private final JwtClaimsSigner jwtClaimsSigner;
+
+    public RestVrpClient(RestOperations restOperations,
+                         JsonConverter jsonConverter,
+                         OAuthClient oAuthClient,
+                         IdempotencyKeyGenerator<OBDomesticVRPConsentRequest, OBDomesticVRPRequest> idempotencyKeyGenerator,
+                         JwtClaimsSigner jwtClaimsSigner) {
+        super(restOperations, jsonConverter, oAuthClient);
+        this.idempotencyKeyGenerator = idempotencyKeyGenerator;
+        this.jwtClaimsSigner = jwtClaimsSigner;
+    }
+
+    @Override
+    public OBDomesticVRPResponse submitDomesticVrp(
+        OBDomesticVRPRequest vrpRequest,
+        AuthorizationContext authorizationContext,
```

Comment:
A VRP consent is a long lived consent and the payment instruction(s) may not happen immediately following the authorization of the consent by the PSU with the ASPSP. This appears to tie the exchange of the auth code to the making of a payment

Could it be better to have the ability to:

- Exchange auth code for access token and refresh token using authorization_code grant
- Exchange refresh token for access token and refresh token using refresh_token grant, if lifetime of access token is limited by the ASPSP and refresh tokens are being used
- Submit a domestic VRP using access token from one of the above 2 methods

This will allow the client application to retrieve the access token within the lifetime of the auth code, make multiple payments for the life of a single access token, and retrieve further access tokens using refresh tokens if the ASPSP supports this

Reply:
Thank you for the your comment! These are valid points. 
Our implementation is one of many possible options. We built openbanking-client in this way because we expect clients to cache tokens on their side.
As an example you can extend `OAuthClient` so it will check cache before requesting a new token. 
_Example:_
 ```
import com.transferwise.openbanking.client.oauth.OAuthClient;
import org.springframework.cache.Cache;

public class CachingOAuthClient implements OAuthClient {

    private final OAuthClient oAuthClient;
    private final Cache accessTokenCache;
    ...
     
    @Override
    public AccessTokenResponse getAccessToken(GetAccessTokenRequest getAccessTokenRequest, AspspDetails aspspDetails) {
        Object someCacheKey = ....
        return accessTokenCache.get(someCacheKey, () -> oAuthClient.getAccessToken(getAccessTokenRequest, aspspDetails));
    }
}
 ```